homeassistant/components/matter/__init__.py

```diff
@@ -94,27 +101,59 @@ async def on_hass_stop(event: Event) -> None:
         hass.data[DOMAIN] = {}
         _async_init_services(hass)
 
-    # we create an intermediate layer (adapter) which keeps track of our nodes
-    # and discovery of platform entities from the node's attributes
+    # create an intermediate layer (adapter) which keeps track of the nodes
+    # and discovery of platform entities from the node attributes
     matter = MatterAdapter(hass, matter_client, entry)
-    hass.data[DOMAIN][entry.entry_id] = matter
+    hass.data[DOMAIN][entry.entry_id] = MatterEntryData(matter, listen_task)
 
-    # forward platform setup to all platforms in the discovery schema
     await hass.config_entries.async_forward_entry_setups(entry, DEVICE_PLATFORM)
+    await matter.setup_nodes()
 
-    # start discovering of node entities as task
-    asyncio.create_task(matter.setup_nodes())
+    # If the listen task is already failed, we need to raise ConfigEntryNotReady
+    await asyncio.sleep(0)  # yield to the listen task at least once again
```

Comment:
How reliable is that in practice :sweat_smile: 

Ideally, I feel we should split up `start_listening` in two commands, one which sets up the listening, and one which enters the while true. This way we could set a flag once initialization is complete and properly wait for that.

Reply:
It is already working like that.

Reply:
Oh right, there is the `init_ready` event. Why is this additional yield needed then?

Reply:
task scheduling

Reply:
@agners are you ok with this now?

Reply:
I still don't understand why the `asyncio.sleep(0)` is needed.

I understand that it essentially kicks the event loop, so making any other task executing.

But this type of manually "kicking the even loop" is usually not good practice and can lead to race conditions.

If the `init_ready` even is all what is needed to start the listening, then why do we need to schedule the event loop again here? What happens if this line is removed?

Reply:
If the server websocket fails after the listen ready event is set, but before we are finished with setting up the config entry, we want to make sure the task has a chance to react to the failure before we do the check here. That's why we yield explicitly to the loop.

Reply:
exactly, the whole point is to prevent a race condition

Reply:
On the other hand, maybe we can skip it, since if the task reacts after the config entry is loaded, it will just reload the config entry which is also ok.

Reply:
both is fine imo

Reply:
I removed it now. The code should already handle both cases that can happen if the websocket fails after the listen ready event is set:

1. Failure where the task has spun once before we check for the exception here.
2. Failure where the task spins after config entry is loaded.

Reply:
> exactly, the whole point is to prevent a race condition

Yes I understand. But it prevents the race condition in that particular (test) setup.

Run it on a slower Raspberry Pi 3 or something, and behavior might change.

> If the server websocket fails after the listen ready event is set, but before we are finished with setting up the config entry, we want to make sure the task has a chance to react to the failure before we do the check here. That's why we yield explicitly to the loop.

If we are concerned about that, an additional event after the first loop or something might be appropriate.

Reply:
It is fine like this. Merge it and move on.